packages/@react-stately/calendar/src/useRangeCalendarState.ts

```diff
@@ -45,17 +45,37 @@ export function useRangeCalendarState<T extends DateValue>(props: RangeCalendarS
     }
   }
 
+  let availableRangeRef = useRef<RangeValue<DateValue>>(null);
+  let availableRange = availableRangeRef.current;
+  let min = useMemo(() => maxDate(minValue, availableRange?.start), [minValue, availableRange]);
+  let max = useMemo(() => minDate(maxValue, availableRange?.end), [maxValue, availableRange]);
+
   let calendar = useCalendarState({
     ...calendarProps,
     value: value && value.start,
     createCalendar,
     locale,
     visibleDuration,
-    minValue,
-    maxValue,
+    minValue: min,
+    maxValue: max,
     selectionAlignment: alignment
   });
 
+  let setAnchorDate = (date: CalendarDate) => {
+    if (date) {
+      setAnchorDateState(date);
+      if (props.isDateUnavailable && !props.allowsNonContiguousRanges) {
+        availableRangeRef.current = {
+          start: nextUnavailableDate(date, calendar, -1),
+          end: nextUnavailableDate(date, calendar, 1)
+        };
```

Comment:
I think I'm reading this right that it'll check for nextUnavailableDate in the current visible month?
What happens if the first date on the next month is unavailable? should that also disable the prev/next month buttons?

Reply:
good point. we should probably look an additional day on either side of the visible range too.

Reply:
actually, that case seems to work fine due to this:

https://github.com/adobe/react-spectrum/blob/7003fe2edc6d95b4426f24293f17138305fd82e0/packages/%40react-stately/calendar/src/useRangeCalendarState.ts#L160-L162

That'll check the first date in the next month, or the last date in the previous month. I added tests for this.

Also fixed some other cases where the available range needs to update when navigating.